### DISCUSSION TITLE
glTF 2.0 specification coverage of gltfio library

### DISCUSSION THREAD

**Discussion by miinso** [google/filament#3401] [General]
I understand glTF is the format used primarily with the filament engine and I notice filament utilizing the cgltf library.

Some of following items are a duplicate or a mere variant of another, but yes there are lots of implementations of the glTF specification.

- [jkuhlmann/cgltf](https://github.com/jkuhlmann/cgltf)
- [KhronosGroup/glTF-Blender-IO](https://github.com/KhronosGroup/glTF-Blender-IO)
- [AnalyticalGraphicsInc/gltf-vscode](https://github.com/AnalyticalGraphicsInc/gltf-vscode)
- [BabylonJS/Babylon.js](https://github.com/BabylonJS/Babylon.js)
- [mrdoob/three.js](https://github.com/mrdoob/three.js)
- [CesiumGS/cesium](https://github.com/CesiumGS/cesium)
- [syoyo/tinygltf](https://github.com/syoyo/tinygltf)
- [jessey-git/fx-gltf](https://github.com/jessey-git/fx-gltf)
- [Microsoft/glTF-SDK](https://github.com/Microsoft/glTF-SDK)
- [Maybe more?](https://user-images.githubusercontent.com/46559594/103451418-499bf700-4d07-11eb-8561-bf8fc4b98efc.jpg)

Each has different compliance and implementation of the glTF spec and their compatible draco version vary.
Most of the time they provide specification chart of their own implementation for the glTF spec, some does not.

Some specs are implemented, some are not.

Every time something doesn't work as expected, it's a whole agonizing _heuristic_ approach again.
I think filament deserve it's own specification chart when it comes to glTF asset.
It will be great ease for us developers to lessen the burden of *Run>Crash>Cryptic tombstone logs* cycles.

1. Does the gltfio library use the cgltf as is, or a subset/variant of?
2. Does cgltf core implementation vary around the supported platforms?
3. Is it safe to assume the gitfio library shares support for the glTF 2.0/Draco in regard to following third party libraries?
    - https://github.com/google/filament/tree/main/third_party/cgltf
    - https://github.com/google/filament/tree/main/third_party/draco

I present my version of checklist anyhow, please check through it, it is worth noting.

Thanks.

**Comment by miinso** [comment]
# gltfio: Features


- Encodings
    - [x] Embeded
    - [x] Binary


- Primitive Types
    - [x] Points
    - [x] Lines
    - [ ] Line Loop
    - [ ] Line Strip
    - [x] Triangles
    - [ ] Triangle Strip
    - [ ] Triangle Fan


- Animation
    - [x] Transform animation
    - [x] Linear interpolation
    - [x] Morph animation
        - [x] Sparse accessor
    - [x] Skin animation
    - [x] Joint animation
    - [x] [google/draco](https://github.com/google/draco) compressed animation


- Extensions 
    - [x] KHR_draco_mesh_compression
        - Compatible version of [google/draco](https://github.com/google/draco) release?
    - [x] KHR_lights_punctual
    - [x] KHR_materials_clearcoat
    - [x] KHR_materials_pbrSpecularGlossiness
    - [x] KHR_materials_sheen
    - [x] KHR_materials_transmission
    - [x] KHR_materials_unlit
    - [ ] KHR_materials_variants
    - [x] KHR_mesh_quantization
    - [ ] KHR_texture_basisu
    - [x] KHR_texture_transform
    - [ ] KHR_xmp
    - [ ] EXT_lights_image_based
    - [ ] EXT_mesh_gpu_instancing
    - [ ] EXT_meshopt_compression
    - [ ] EXT_texture_webp

**Comment by romainguy** [comment]
gltfio offers a full implementation of the core glTF 2.0 spec. In addition the following extensions are supported:

  KHR_draco_mesh_compression
  KHR_lights_punctual
  KHR_materials_clearcoat
  KHR_materials_pbrSpecularGlossiness
  KHR_materials_sheen
  KHR_materials_transmission
  KHR_materials_unlit
  KHR_texture_transform

KHR_materials_ior and _specular should be added soon as well.

We do not attempt to support extensions that are not ratified or not close to being ratified (basically we only look at KHR_ extensions).

Also not that we not only support the latest cgltf but we actively contribute to it.

**Comment by cx20** [comment]
@61315 For reference, I'm investigating glTF support in the [gltf-test](https://github.com/cx20/gltf-test) repository for some WebGL libraries.

**Comment by prideout** [comment]
Romain's response is correct.  We do not support loops, strips, or fans but otherwise we should be fully conformant.  I've edited the checklist appropriately.

**Comment by cx20** [comment]
I tested [the model](https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/Duck/glTF-Quantized) of the [KHR_mesh_quantization](https://github.com/KhronosGroup/glTF/tree/master/extensions/2.0/Khronos/KHR_mesh_quantization) extension with [gltf-test](https://github.com/cx20/gltf-test).
`Filament` can display that model, so it probably supports the `KHR_mesh_quantization` extension.

https://cx20.github.io/gltf-test/examples/filament/index.html?model=Duck&scale=1&type=glTF-Quantized

**Comment by romainguy** [comment]
It looks like we do, indirectly, thanks to how cgltf handles it.

**Comment by romainguy** [comment]
I'm going to add this list of supported glTF features to our main README btw

**Comment by issacclee** [comment]
#1378 
It is mentioned in this issue that support of KHR_texture_basisu is under consideration. I was wondering if there's any decision has been made regarding to whether or not filament will support this extension in the future?

**Reply by romainguy** [reply]
Filament will support it eventually but I don't know when.

**Reply by issacclee** [reply]
> Filament will support it eventually but I don't know when.

Got it, Thanks!

**Comment by PsycoTodd** [comment]
Hello, 

I am working on parsing the [KHR_xmp_json_ld](https://github.com/KhronosGroup/glTF/blob/main/extensions/2.0/Khronos/KHR_xmp_json_ld/README.md) extension in filament. at cgltf.h level. 

But I have a question as I can get the data but decide to give the json as data to native side. Since they are just meta data and native (java/ios/web) can decide how to use them. 

However, I only see `extras` has past up to AssetLoader level. 

My question is for these extensions that do not change the appearance of the render result, would filament consider to have way to let native side know? One case is: how would a java code know the copy right information for a loaded asset (gltf asset.copyright). 

Thank you.

**Reply by romainguy** [reply]
`extras` are the generic way to pass data up. Otherwise we (or you) would have to add support for this specific extension.

**Reply by PsycoTodd** [reply]
yeah, working on it following the extra examples. will share. thank you
